<a href="https://colab.research.google.com/github/omaskii/Complete-Python-3-Bootcamp/blob/master/NumPy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Solomon Akatakpo** - Lecture notes from Great Learning

In [15]:
# The first step of using numpy is to tell python to import it
import numpy as np

An array is a data structure that stores values of same data type.
While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to the same data type.
However python lists fail to deliver the performance required while computing large sets of numerical data. To address this issue we use NumPy arrays.
We can create NumPy arrays by converting a list to an array.

In [2]:
# defining a list of different car companies or string elements
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 3]

In [3]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)


# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  [5 4 6 7 3]


The resuts look similar to a list but arr_str and arr_num have been converted to NumPy arrays. Let's check the data type to confirm this.

In [4]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


The above output confirms that both the lists were successfully converted to arrays

NumPy Matrix
A matrix is a two-dimensional data structure where elements are arranged into rows and columns.
A matrix can be created by using list of lists

In [5]:
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [6]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


We see that all the NumPy objects have data type as ndarray

There are different ways to create NumPy arrays using the functions available in NumPy library

Using np.arange() function

The np.arange() function returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included but stop is excluded.
It has the following paramaters:
start : start of interval range. By default start = 0
stop : end of interval range
step : step size of interval. By default step size = 1

In [7]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

# or

arr2  = np.arange(0,10) 
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [8]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 20, step = 5)
arr3

array([ 0,  5, 10, 15])

In [9]:
np.arange(0,20,5)

array([ 0,  5, 10, 15])

Using np.linspace() function

The np.linspace() function returns numbers which are evenly distributed with respect to interval. Here the start and stop both are included.
*It has the following parameters:
start: start of interval range. By default start = 0
stop: end of interval range
num : No. of samples to generate. By default num = 50

In [10]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

How are these values getting generated?

The step size or the difference between each element will be decided by the following formula:

(stop - start) / (total elements - 1)

So, in this case: (5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on.

In [11]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,10)
matrix3

array([10.        , 11.11111111, 12.22222222, 13.33333333, 14.44444444,
       15.55555556, 16.66666667, 17.77777778, 18.88888889, 20.        ])

Similarly we can create matrices using the functions available in NumPy library

Using np.zeros()

The np.zeros() is a function for creating a matrix and performing matrix operations in NumPy.
It returns a matrix filled with zeros of the given shape.
It has the following parameters:
shape : Number of rows and columns in the output matrix.
dtype: data type of the elements in the matrix, by default the value is set to float.

In [12]:
matrix4 = np.zeros([3,5])
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

Using np.ones()

The np.ones() is another function for creating a matrix and performing matrix operations in NumPy.
It returns a matrix of given shape and type, filled with ones.
It has the following parameters:
shape : Number of rows and columns in the output matrix.
dtype: data type of the elements in the matrix, by default the value is set to float.

In [14]:
Matrix_ones = np.ones([5,10])
Matrix_ones

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

Using np.eye()

The np.eye() is a function for creating a matrix and performing matrix operations in NumPy.
It returns a matrix with ones on the diagonal and zeros elsewhere.
It has the following parameters:
n: Number of rows and columns in the output matrix
dtype: data type of the elements in the matrix, by default the value is set to float.

In [16]:
Matrix6 = np.eye(4)
Matrix6

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

We can also convert a one dimension array to a matrix. This can be done by using the np.reshape() function.

The shape of an array basically tells the number of elements and dimensions of the array. Reshaping a Numpy array simply means changing the shape of the given array.
By reshaping an array we can add or remove dimensions or change number of elements in each dimension.
In order to reshape a NumPy array, we use the reshape method with the given array.
Syntax: array.reshape(shape)
shape: a tuple given as input, the values in tuple will be the new shape of the array.

In [21]:
# defining an array with values 0 to 14
arr4 = np.arange(0,15) 
arr4

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [22]:
# the way to reshape the above one dimensional array into a 2 dimensional array is given below
arr4_reshaped = arr4.reshape((3,5))
arr4_reshaped

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

NumPy can also perform a large number of different mathematical operations and it provides different functions to do so.

NumPy provides:

Trigonometric functions
Exponents and Logarithmic functions
Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [23]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495775


**Exponents and Logarithmic functions**

Exponents

In [24]:
np.exp(2)

7.38905609893065

In [25]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

Logarithms

In [26]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [27]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [28]:
## log with base 10
np.log10(8) 

0.9030899869919435

Arithmetic Operations on arrays

In [29]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
# this does not behave as you would expect!

[1, 2, 3, 4, 5, 6]


In [30]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [31]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!


Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [32]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [33]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


<ipython-input-33-b49c47b4c85a>:4: RuntimeWarning: divide by zero encountered in true_divide
  print('Division: \n', matrix7/matrix8)


RuntimeWarning: Errors which occur during program execution(run-time) after successful compilation are called run-time errors.
One of the most common run-time error is division by zero also known as Division error.
Due to division by zero error, we are getting inf (infinity) values because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [34]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


**Transpose of a matrix**

In [35]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [36]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [37]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [38]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [39]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [40]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the normal distribution over [0,1). (including 0 but excluding 1).
Syntax - np.random.rand(d0,d1)
d0,d1 – It represents the dimension of the required array given as int, where d1 is optional.

In [42]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.07819633 0.34821932 0.21733969 0.80823843 0.98242399]


In [43]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.13960542 0.45540405 0.21548702 0.0442641  0.02446824]
 [0.30794651 0.41633487 0.20223996 0.95331521 0.81797255]
 [0.91042531 0.10502654 0.34020142 0.21724663 0.41536841]
 [0.94150843 0.53527046 0.03777765 0.27651013 0.03143651]
 [0.46349293 0.40629505 0.9446929  0.79763601 0.76390177]]
